src/emqx_router.erl

```diff
@@ -248,46 +281,47 @@ delete_trie_route(Route = #route{topic = Topic}) ->
     end.
 
 %% @private
--spec(maybe_trans(function(), list(any())) -> ok | {error, term()}).
-maybe_trans(Fun, Args) ->
-    case persistent_term:get(emqx_route_lock_type) of
+-spec(maybe_trans(pid(), function(), list(any())) -> ok | {error, term()}).
+maybe_trans(Worker, Fun, Args) ->
+    case persistent_term:get(emqx_route_lock_type, key) of
         key ->
-            trans(Fun, Args);
+            trans(Worker, Fun, Args);
         global ->
             lock_router(),
             try mnesia:sync_dirty(Fun, Args)
             after
                 unlock_router()
             end;
         tab ->
-            trans(fun() ->
+            trans(Worker,
+                  fun() ->
                           emqx_trie:lock_tables(),
                           apply(Fun, Args)
                   end, [])
     end.
 
--spec(trans(function(), list(any())) -> ok | {error, term()}).
-trans(Fun, Args) ->
+-spec(trans(pid(), function(), list(any())) -> ok | {error, term()}).
+trans(Worker, Fun, Args) when is_pid(Worker) ->
     %% trigger selective receive optimization of compiler,
     %% ideal for handling bursty traffic.
-    Ref = erlang:make_ref(),
-    Owner = self(),
-    {WPid, RefMon} = spawn_monitor(
-                       fun() ->
-                               Res = case mnesia:transaction(Fun, Args) of
-                                         {atomic, Ok} -> Ok;
-                                         {aborted, Reason} -> {error, Reason}
-                                     end,
-                               Owner ! {Ref, Res}
-                       end),
+    MonRef = erlang:monitor(process, Worker),
+    Worker ! {work, MonRef, Fun, Args},
     receive
-        {Ref, TransRes} ->
-            receive
-                {'DOWN', RefMon, process, WPid, normal} -> ok
-            end,
+        {MonRef, TransRes} ->
+            case erlang:demonitor(MonRef, [info]) of
```

Comment:
this is the same as `erlang:demonitor(MonRef, [flush])` in OTP internals, I will revert this change.